source/common/network/connection_impl.cc

```diff
@@ -83,7 +83,10 @@ ConnectionImpl::ConnectionImpl(Event::Dispatcher& dispatcher, ConnectionSocketPt
       transport_wants_read_(false) {
 
   if (!connected) {
+    // immediate_error_event_ is nullopt here and event will be ignored.
     connecting_ = true;
+  } else {
+    immediate_error_event_ = ConnectionEvent::Connected;
```

Comment:
immediate_error_event_ is not meaningful when it is set to Connected.  Connected is being used as a sentinel.  You shouldn't need to ever set immediate_error_event_ to Connected.

Reply:
Yes, I understand Connected is a sentinel. My major point is that we change the Connected as sentinel to nullopt as sentinel. 
From
`ConnectionEvent immediate_error_event_{ConnectionEvent::Connected};`
to
`absl::optional<ConnectionEvent> immediate_error_event_;`